src/stash/src/upgrade.rs

```diff
@@ -23,32 +33,75 @@ pub enum MigrationAction<K1, K2, V2> {
     Update(K1, (K2, V2)),
 }
 
-/// TODO(parkmycar): Once the Stash / the [`crate::Data`] trait uses [`prost`], get rid of this.
-pub trait DataProto: ::prost::Message + std::cmp::Ord + Default + Send + Sync {}
-impl<T: ::prost::Message + std::cmp::Ord + Default + Send + Sync> DataProto for T {}
-
 impl<K, V> TypedCollection<K, V>
 where
-    K: DataProto,
-    V: DataProto,
+    K: Data,
+    V: Data,
 {
     /// Provided a closure, will migrate a [`TypedCollection`] of types `K` and `V` to types `K2`
     /// and `V2`.
-    ///
-    /// TODO(parkmycar): Fill in this implementation in a future PR.
-    #[allow(clippy::unused_async)]
     pub async fn migrate_to<K2, V2>(
         &self,
-        _tx: &mut Transaction<'_>,
-        _f: impl for<'a> FnOnce(&'a BTreeMap<K, V>) -> Vec<MigrationAction<K, K2, V2>>,
+        tx: &mut Transaction<'_>,
+        f: impl for<'a> FnOnce(&'a BTreeMap<K, V>) -> Vec<MigrationAction<K, K2, V2>>,
     ) -> Result<(), StashError>
     where
-        K2: DataProto,
-        V2: DataProto,
+        K2: Data,
+        V2: Data,
     {
-        Err(StashError {
-            inner: InternalStashError::Other("unimplemented".to_string()),
-        })
+        // Create a batch that we'll write to.
+        let collection = tx.collection::<K, V>(self.name).await?;
+        let lower = tx.upper(collection.id).await?;
+        let mut batch = collection.make_batch_lower(lower)?;
+        let current = match tx.peek_one(collection).await {
+            Ok(set) => set,
+            Err(err) => match err.inner {
+                InternalStashError::PeekSinceUpper(_) => {
+                    // If the upper isn't > since, bump the upper and try again to find a sealed
+                    // entry. Do this by appending the empty batch which will advance the upper.
+                    tx.append(vec![batch]).await?;
+                    let lower = tx.upper(collection.id).await?;
+                    batch = collection.make_batch_lower(lower)?;
+                    tx.peek_one(collection).await?
+                }
+                _ => return Err(err),
+            },
+        };
+
+        // Note: this method exists, instead of using `StashCollection::append_to_batch` so we can
+        // append types other than K or V.
+        fn append_to_batch<A, B>(batch: &mut AppendBatch, k: &A, v: &B, diff: i64)
+        where
+            A: ::prost::Message,
+            B: ::prost::Message,
+        {
+            let key = k.encode_to_vec();
+            let val = v.encode_to_vec();
+
+            batch.entries.push(((key, val), batch.timestamp, diff));
+        }
+
+        // Call the provided closure, generating a list of update actions.
+        for op in f(&current) {
+            match op {
+                MigrationAction::Delete(old_key) => {
```

Comment:
The Delete block appears not to be tested

Reply:
No migrations need to delete keys yet, since this PR is so large do you mind if I follow up with this?